### PR TITLE
Feat/merkle tree build

### PR DESCRIPTION
**Pull request by LucasUTNFRD**
Add basic implementation and test for the core build tree function

### CODE REVIEW

File: src/merkle.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
+use sha3::{Digest, Sha3_256};
+
+type Hash = Vec<u8>;
```
Comment (LeanSerra) [inline src/merkle.rs]: Hashes are fixed length, you can use [u8; 32]
```suggestion
type Hash = [u8; 32];
```

File: src/merkle.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    hasher.update(right);
+    hasher.finalize().to_vec()
+}
+
+impl Merkle {
+    /// Creates a new Merkle tree from a list of data elements.
+    /// TODO: Make this function generic to accept any type that can be hashed.
+    ///
+    pub fn new(data: &[Vec<u8>]) -> Self {
+        assert!(data.len() > 1, "Data must have at least 2 elements");
```
Comment (LeanSerra) [inline src/merkle.rs]: You can create a merkle tree with only one element, the result is a tree with only one element, that element being the root

File: src/merkle.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            while i < count_leaves {
+                if i + 1 < count_leaves {
+                    // Combine adjacent hashes
+                    level_hashes.push(hash_internal_node(&current_level[i], &current_level[i + 1]));
+                } else {
+                    // Clone the last hash to create a pair
+                    level_hashes.push(hash_internal_node(&current_level[i], &current_level[i]));
+                }
+                i += 2;
+            }
```
Comment (LeanSerra) [inline src/merkle.rs]: Using a for loop with a range could be more idiomatic
```suggestion
            for i in (0..count_leaves).step_by(2) {
                if i + 1 < count_leaves {
                    // Combine adjacent hashes
                    level_hashes.push(hash_internal_node(&current_level[i], &current_level[i + 1]));
                } else {
                    // Clone the last hash to create a pair
                    level_hashes.push(hash_internal_node(&current_level[i], &current_level[i]));
            }
        }
```

**Review by LeanSerra — CHANGES_REQUESTED** [review]
Left some suggestions, but looks good.

**Review by damiramirez — APPROVED** [review]
LGTM + Lean comments

File: src/merkle.rs (hunk 1 above)
Comment (LucasUTNFRD) [reply src/merkle.rs]: Done

File: src/merkle.rs (hunk 2 above)
Comment (LucasUTNFRD) [reply src/merkle.rs]: Done

File: src/merkle.rs (hunk 3 above)
Comment (LucasUTNFRD) [reply src/merkle.rs]: Done

File: src/merkle.rs [hunk 4]
```diff
@@ -1,29 +1,29 @@
...
+    hasher.finalize().into()
 }
 
 impl Merkle {
     /// Creates a new Merkle tree from a list of data elements.
     /// TODO: Make this function generic to accept any type that can be hashed.
     ///
     pub fn new(data: &[Vec<u8>]) -> Self {
-        assert!(data.len() > 1, "Data must have at least 2 elements");
+        assert!(!data.is_empty(), "Data must have at least 1 elements");
```
Comment (LeanSerra) [inline src/merkle.rs:26]: I would return an error if the array is empty.

**Review by LeanSerra — APPROVED** [review]
LGTM, the comment I made can be addressed in the next PR

**Review by damiramirez — APPROVED** [review]

File: src/merkle.rs (hunk 4 above)
Comment (LucasUTNFRD) [reply src/merkle.rs:26]: I will add errors in next PR, is it ok?

**Pull request merged**